File: src/clu/sources/uptime.py
import logging
import re

from clu.provides import Provides
from clu.requires import Requires
from clu.facts import Facts
from clu.sources import Source
from clu.input import text_program
from clu.sources import PARSE_FAIL_MSG

log = logging.getLogger(__name__)


class Uptime(Source):
    def provides(self, provides: Provides) -> None:
        provides["run.uptime"] = self

    def requires(self, requires: Requires) -> None:
        requires.programs.append("uptime")

    def parse(self, facts: Facts) -> None:
        data, rc = text_program("uptime")
        log.debug(f"{data=}")
        if data == "" or rc != 0:
            facts["run.uptime"] = PARSE_FAIL_MSG
            return

        match = re.match(r".*up *(.*) \d+ user.*", data)
        uptime = match.group(1).rstrip(",") if match else PARSE_FAIL_MSG
        log.debug(f"{uptime=}")
        facts["run.uptime"] = uptime

File: pkg/facts/sources/uptime.go
package sources

import (
	"regexp"
	"strings"

	"github.com/NHGRI/clu/pkg/facts/types"
	"github.com/NHGRI/clu/pkg/input"
)

// Uptime parses system uptime via `uptime` command.
type Uptime struct{}

var uptimeFactCmd = types.Fact{
	Name: "run.uptime",
	Tier: types.TierOne,
}

func (u *Uptime) Provides(p types.Provides) {
	p[uptimeFactCmd.Name] = u
}

func (u *Uptime) Requires(r *types.Requires) {
	r.Programs = append(r.Programs, "uptime")
}

// Match both singular 'user' and plural 'users'.
var uptimeRegex = regexp.MustCompile(`.*up *(.*) \d+ user(?:s)?,? .*`)

func (u *Uptime) Parse(f *types.FactDB) {
	data, rc, _ := input.CommandRunner("uptime")
	if data == "" || rc != 0 {
		uptimeFactCmd.Value = types.ParseFailMsg
		f.AddFact(uptimeFactCmd)
		return
	}

	m := uptimeRegex.FindStringSubmatch(data)
	if len(m) > 1 {
		uptime := strings.TrimSuffix(m[1], ",")
		uptimeFactCmd.Value = strings.Join(strings.Fields(uptime), " ")
	} else {
		uptimeFactCmd.Value = types.ParseFailMsg
	}
	f.AddFact(uptimeFactCmd)
}
